lib/pengine/unpack.c

```diff
@@ -2797,6 +2797,60 @@ newer_state_after_migrate(const char *rsc_id, const char *node_name,
                                         data_set);
 }
 
+/*!
+ * \internal
+ * \brief Parse migration source and target node names from history entry
+ *
+ * \param[in] entry         Resource history entry for a migration action
+ * \param[in] source_node   If not NULL, source must match this node
+ * \param[in] target_node   If not NULL, target must match this node
+ * \param[out] source_name  Where to store migration source node name
+ * \param[out] target_name  Where to store migration target node name
+ *
+ * \return Standard Pacemaker return code
+ */
+static int
+get_migration_node_names(const xmlNode *entry, const pe_node_t *source_node,
+                         const pe_node_t *target_node,
+                         const char **source_name, const char **target_name)
+{
+    const char *id = ID(entry);
+
+    if (id == NULL) {
+        crm_err("Ignoring resource history entry without ID");
+        return pcmk_rc_unpack_error;
+    }
+
+    *source_name = crm_element_value(entry, XML_LRM_ATTR_MIGRATE_SOURCE);
+    *target_name = crm_element_value(entry, XML_LRM_ATTR_MIGRATE_TARGET);
+    if ((*source_name == NULL) || (*target_name == NULL)) {
+        crm_err("Ignoring resource history entry %s without "
+                XML_LRM_ATTR_MIGRATE_SOURCE " and " XML_LRM_ATTR_MIGRATE_TARGET,
+                id);
+        return pcmk_rc_unpack_error;
+    }
+
+    if ((source_node != NULL)
+        && !pcmk__str_eq(*source_name, source_node->details->uname,
+                         pcmk__str_casei|pcmk__str_null_matches)) {
```

Comment:
Just a bit of future-proofing. With the current code, we should never have a node without a name here, but if we ever did, it would mean we know what node we're unpacking history for but not its name. In that case it would be better to assume that the resource history entry matches the node state it's in than otherwise.